### PR TITLE
RM-7663: Remove jQuery usages from BocList

### PR DESCRIPTION
**Pull request by Stefan-Ilic**
https://re-motion.atlassian.net/browse/RM-7663

### CODE REVIEW

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 1]
```diff
@@ -395,35 +399,37 @@ class BocList
...
+    bocList.parentNode!.insertBefore(tempList, bocList)
   
     try
     {
-      if (tempList.height() > referenceHeight)
+      if (BocList.getComputedHeight(tempList) > referenceHeight)
         return true;
   
-      if (tempList.width() > referenceWidth)
+      if (BocList.getComputedWidth(tempList) > referenceWidth)
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:432]: All `.height` calls should be converted to calls to the `LayoutHelper`.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 2]
```diff
@@ -432,14 +438,14 @@ class BocList
...
+    bocList.parentNode!.insertBefore(tempList, bocList.nextSibling)
   
     try
     {
-      if (tempList.height() > referenceHeight)
+      if (tempList.clientHeight > referenceHeight)
         return true;
   
-      if (tempList.width() > referenceWidth)
+      if (tempList.clientWidth > referenceWidth)
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:448]: Also covert to `LayoutHelper` calls

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 3]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    fakeTableHeadContainer.style.display = 'none';
     fakeTableHeadContainer.append (fakeTableHeadWidthContainer);
   
-    realTableHead.find('*').each(function (this: HTMLElement) { $(this).removeAttr('id').attr({ tabIndex: -1 }).attr({ tabIndex: -1 }); });
+
+    realTableHead.querySelectorAll('*').forEach(element =>
+    {
+      element.removeAttribute('id');
+      element.setAttribute('tabIndex', '-1');
+    });
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:554]:
```suggestion
    realTableHead.querySelectorAll('*').forEach(element =>
    {
      element.removeAttribute('id');
      element.setAttribute('tabIndex', '-1');
    });
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 4]
```diff
@@ -492,116 +498,125 @@ class BocList
...
-      $('input[name="' + checkName + '"]').prop('checked', checkStatus);
+      checkbox.addEventListener('click', ((event: MouseEvent) =>
+      {
+        var checkName = checkbox.getAttribute('name');
+        var checkStatus = (checkbox as HTMLInputElement).checked
+        document.querySelectorAll('input[name="' + checkName + '"]').forEach(element =>
+        {
+          (element as HTMLInputElement).checked = checkStatus
+        });
+      }) as EventListener)
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:570]: Remove the cast to `as EventListener` and the explicit parameter type annotation.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 5]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+      element.removeAttribute('id');
+      element.setAttribute('tabIndex', '-1');
+    });
   
     scrollableContainer.before(fakeTableHeadContainer);
   
     // sync checkboxes
-    var checkboxes = fakeTableHead.find("th input:checkbox");
-    checkboxes.click(function (this: HTMLElement)
+    var checkboxes = fakeTableHead.querySelectorAll("th input[type=checkbox]");
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:559]: We should probably use `<HTMLInputElement>` here so we can remove following cast

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 6]
```diff
@@ -492,116 +498,125 @@ class BocList
...
-      var checkName = $(this).attr('name');
-      var checkStatus = $(this).prop('checked');
-      $('input[name="' + checkName + '"]').prop('checked', checkStatus);
+      checkbox.addEventListener('click', ((event: MouseEvent) =>
+      {
+        var checkName = checkbox.getAttribute('name');
+        var checkStatus = (checkbox as HTMLInputElement).checked
+        document.querySelectorAll('input[name="' + checkName + '"]').forEach(element =>
+        {
+          (element as HTMLInputElement).checked = checkStatus
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:568]:
```suggestion
          (element as HTMLInputElement).checked = checkStatus;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 7]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    var checkboxes = fakeTableHead.querySelectorAll("th input[type=checkbox]");
+    checkboxes.forEach(checkbox =>
     {
-      var checkName = $(this).attr('name');
-      var checkStatus = $(this).prop('checked');
-      $('input[name="' + checkName + '"]').prop('checked', checkStatus);
+      checkbox.addEventListener('click', ((event: MouseEvent) =>
+      {
+        var checkName = checkbox.getAttribute('name');
+        var checkStatus = (checkbox as HTMLInputElement).checked
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:565]:
```suggestion
        var checkStatus = (checkbox as HTMLInputElement).checked;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 8]
```diff
@@ -492,116 +498,125 @@ class BocList
...
   }
 
-  private static FixHeaderSize(scrollableContainer: JQuery): void
+  private static FixHeaderSize(scrollableContainer: HTMLElement): void
   {
-    var realTable = scrollableContainer.children('table').first();
-    var realTableWidth = realTable[0].offsetWidth;
-    var previousRealTableWidth = realTable.data("bocListPreviousRealTableWidth");
+    var realTable = scrollableContainer.querySelector<HTMLTableElement>(':scope>table')!;
+    var realTableWidth = BocList.getComputedWidth(realTable);
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:578]: This should stay `.offsetWidth`

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 9]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    var previousRealTableWidth = parseInt(realTable.dataset.bocListPreviousRealTableWidth!)
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:579]:
```suggestion
    var previousRealTableWidth = parseInt(realTable.dataset.bocListPreviousRealTableWidth!);
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 10]
```diff
@@ -492,116 +498,125 @@ class BocList
...
     if (previousRealTableWidth == realTableWidth)
       return;
-    realTable.data("bocListPreviousRealTableWidth", realTableWidth);
+    realTable.dataset.bocListPreviousRealTableWidth = realTableWidth.toString()
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:582]:
```suggestion
    realTable.dataset.bocListPreviousRealTableWidth = realTableWidth.toString();
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 11]
```diff
@@ -492,116 +498,125 @@ class BocList
...
     });
   
     // apply widths to fake header
-    fakeTableHeadRowChildren.width(function (index, itemWidth)
+    Array.from(fakeTableHeadRowChildren).forEach((element: Element, index: number) =>
     {
-      var width = realTableHeadCellWidths[index];
-      return width;
+      (element as HTMLElement).style.width = realTableHeadCellWidths[index] + 'px';
     });
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:604]: If we need any of the array methods we should convert to an array when declaring the variable.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 12]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    var fakeTableHeadRow = fakeTableHead.children[0];
+    var fakeTableHeadRowChildren = fakeTableHeadRow.children;
   
     // store cell widths in array
     var realTableHeadCellWidths = new Array();
-    realTableHeadRowChildren.each(function (this: HTMLElement, index)
+    Array.from(realTableHeadRowChildren).forEach(function (element: Element, index: number)
     {
-      realTableHeadCellWidths[index] = $(this).width();
+      realTableHeadCellWidths[index] = BocList.getComputedWidth(element as HTMLElement);
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:597]: This should use the `LayoutHelper`

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 13]
```diff
@@ -492,116 +498,125 @@ class BocList
...
-    var fakeTableHeadWidthContainer = fakeTableHeadContainer.children ('div').first();
-    fakeTableHeadWidthContainer.width(realTableWidth);
-    var fakeTableHeadContainerHeight = fakeTableHeadContainer.height();
-    scrollableContainer.css({ top: fakeTableHeadContainerHeight});
-    realTable.css({ 'margin-top': fakeTableHeadContainerHeight * -1 });
-  
-    fakeTableHeadContainer.show();
+
+    var fakeTableHeadWidthContainer = fakeTableHeadContainer.querySelector<HTMLElement> (':scope>div')!;
+    fakeTableHeadWidthContainer.style.width = realTableWidth + 'px'
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:607]:
```suggestion
    fakeTableHeadWidthContainer.style.width = realTableWidth + 'px';
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 14]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    fakeTableHeadContainer.style.display = 'block';
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:608]: Why was this moved up?

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 15]
```diff
@@ -492,116 +498,125 @@ class BocList
...
+    scrollableContainer.style.top = fakeTableHeadContainerHeight + 'px';
+    realTable.style.marginTop = (fakeTableHeadContainerHeight * -1) + 'px';
   }
 
-  private static FixHeaderPosition(tableContainer: JQuery, scrollableContainerJQuery: JQuery): void
+  private static FixHeaderPosition(tableContainer: HTMLElement, scrollableContainerJQuery: HTMLElement): void
   {
-    var scrollableContainer = scrollableContainerJQuery[0];
-    var fakeTableHeadContainer = tableContainer.children('div.bocListFakeTableHead').first()[0];
+    var scrollableContainer = scrollableContainerJQuery;
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:616]: We should rename the parameter and remove the variable declaration

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 16]
```diff
@@ -672,4 +687,14 @@ class BocList
...
+
+  private static getComputedWidth(htmlElement: HTMLElement): number
+  {
+    return Math.floor(parseFloat(getComputedStyle(htmlElement).width.replace('px', '')))
+  }
+
+  private static getComputedHeight(htmlElement: HTMLElement): number
+  {
+    return Math.floor(parseFloat(getComputedStyle(htmlElement).height.replace('px', '')))
+  }
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:699]: Those should be removed once the `LayoutHelper` methods are in place.

File: Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocListNavigationBlockRenderer.cs [hunk 17]
```diff
@@ -207,7 +207,7 @@ private void RenderPageInformation (BocListRenderingContext renderingContext)
       renderingContext.Writer.RenderEndTag(); // end span
 
       string script = string.Format (
-          "BocList.InitializeNavigationBlock ($('#{0}'), $('#{1}'))",
+          "BocList.InitializeNavigationBlock (document.getElementById('{0}'), document.getElementById('{1}'))",
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocListNavigationBlockRenderer.cs:210]: `BocList.InitializeNavigationBlock` should use the HTMLElement or CSS selector pattern so that we can use a css selectors here. See handbook, under concept "jQuery objects as parameters"

File: Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocListRenderer.cs [hunk 18]
```diff
@@ -263,7 +263,7 @@ private void RegisterInitializeListScript (BocListRenderingContext renderingCont
         bool hasClickSensitiveRows = renderingContext.Control.IsSelectionEnabled && !renderingContext.Control.EditModeController.IsRowEditModeActive
                                      && renderingContext.Control.AreDataRowsClickSensitive();
 
-        const string scriptTemplate = "BocList.InitializeList ( $('#{0}')[0], '{1}', '{2}', {3}, {4}, {5});";
+        const string scriptTemplate = "BocList.InitializeList ( document.getElementById('{0}'), '{1}', '{2}', {3}, {4}, {5});";
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web/UI/Controls/BocListImplementation/Rendering/BocListRenderer.cs:266]: `BocList.InitializeList` should use the HTMLElement or CSS selector pattern

**Review by a-ctor — CHANGES_REQUESTED** [review]
I have made the mistake of thinking clientHeight is the correct equivalent to jQuery's `.height()`, which it isn't. As such we need to create helpers for it.

Please create a commit adding those two methods to the LayoutHelper and use them instead of `.clientHeight`. Depending on which PR is merged first (yours or Marcels) the commit will be empty on rebase, since the other PR already contained the same code.

```js
public static GetHeight(element: HTMLElement)
{
  // TODO RM-7734: Fix the width/height calculations in the LayoutHelper to respect box-sizing
  const style = window.getComputedStyle(element);
  return parseFloat(style.height) || 0;
}

public static GetWidth(element: HTMLElement)
{
  // TODO RM-7734: Fix the width/height calculations in the LayoutHelper to respect box-sizing
  const style = window.getComputedStyle(element);
  return parseFloat(style.height) || 0;
}
```

`BocList.OnSelectAllSelectorControlClick` should also be converted to the HTML element or CSS selector pattern.

**Comment by a-ctor** [comment]
#63 is now merged and the new `LayoutHelper` methods are available after rebasing.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 14 above)
Comment (Stefan-Ilic) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:608]: Excellent question; `getComputedStyle` always returns 0 for height and width if the element's display style is `none` or `inline`, so it needs to become a `block` before we get its height and width

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 8 above)
Comment (Stefan-Ilic) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:578]: No can do, `ScreenshotTest_TableContainer` fails if I do so

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 19]
```diff
@@ -104,43 +104,44 @@ class BocList
...
 
-      selectedRows.SelectRowSelectorControls.each (function (this: JQuery)
+      selectedRows.SelectRowSelectorControls.forEach (checkBox =>
       {
         selectedRows.DataRowCount++;
 
-        var checkBox = this;
-
-        var tableCell = checkBox.parentNode;
+        var tableCell = checkBox.parentNode!;
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:130]: We should do the `as HTMLElement` cast here as well...

```suggestion
        var tableCell = checkBox.parentNode as HTMLElement;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 20]
```diff
@@ -104,43 +104,44 @@ class BocList
...
           throw 'Unexpected element type: \'' + tableCell.nodeName + '\'';
 
-        var tableRow = tableCell.parentNode;
+        var tableRow = tableCell.parentNode as HTMLElement;
         if (tableRow.nodeName !== 'TR')
           throw 'Unexpected element type: \'' + tableRow.nodeName + '\'';
 
         if (hasClickSensitiveRows)
-          BocList.BindRowClickEventHandler (bocList, tableRow, tableCell, this);
+          BocList.BindRowClickEventHandler (bocList, tableRow as HTMLElement, tableCell as HTMLElement, checkBox);
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:139]: ...and then we can remove the casts from here.

```suggestion
          BocList.BindRowClickEventHandler (bocList, tableRow, tableCell, checkBox);
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 21]
```diff
@@ -314,28 +319,28 @@ class BocList
...
+    selectedRows.SelectAllSelectorControls!.forEach (element => { element.checked = false; }); // TODO RM-7711 - Move BocList's Row Selection logic to TypeScript class 'BocList_SelectedRows'.
   }
 
   //  Event handler for the selection selectorControl in the title row.
   //  Applies the checked state of the title's selectorControl to all data rows' selectu=ion selectorControles.
   //  bocList: The BocList to which the selectorControl belongs.
   //  selectRowControlName: The name of the row selector controls.
   //  isRowHighlightingEnabled: true to enable highting of the rows
-  public static OnSelectAllSelectorControlClick (bocList: HTMLElement, selectAllSelectorControl: JQuery, isRowHighlightingEnabled: boolean): void
+  public static OnSelectAllSelectorControlClick (bocList: HTMLElement, selectAllSelectorControl: HTMLInputElement, isRowHighlightingEnabled: boolean): void
```
Comment (a-ctor) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:343]: Since this method is public we should introduce the selector pattern as well to stay consistent.

**Review by a-ctor — CHANGES_REQUESTED** [review]
I am going to fix the typings for `LayoutUtility` in a separate issue so that we can remove it from this PR.

Update: @Stefan-Ilic The new typings are merged so when your feature branch is rebased the changes to Utility.d.ts should disappear

**Review by a-ctor — APPROVED** [review]

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 22]
```diff
@@ -167,24 +168,28 @@ class BocList
...
+    let prev: Nullable<Node> = cell;
+    while (prev = prev.previousSibling)
     {
-      BocList.OnRowSelectorClick();
-    });
+      prev.addEventListener('click', function (evt)
+      {
+        BocList.OnRowSelectorClick();
+      });
+    }
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:192]: please update to a for-loop. We try to avoid having two effects per statement, here it's assignment and test for null

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 23]
```diff
@@ -167,24 +168,28 @@ class BocList
...
-    $(cell).click(function (evt)
+    cell.addEventListener('click', function (evt)
     {
       BocList.OnRowSelectorClick();
     });
   
     // Enable the entire row to the left of the selector control's cell for click events that mimic the selector control.
-    $(cell).prevAll().click(function (evt)
+    let prev: Nullable<Node> = cell;
+    while (prev = prev.previousSibling)
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:186]: the loop-variable should perhaps not be 'prev' / 'previous' but 'current'. That we iterate the previous items is declared via 'previousSibling', but making 'cell' the 'prev' just feels strange.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 24]
```diff
@@ -432,14 +442,14 @@ class BocList
     }
   
     // Catch styles applied to pseudo-selectors starting at the last element in the DOM collection
-    tempList.insertAfter($(bocList));
+    bocList.parentNode!.insertBefore(tempList, bocList.nextSibling)
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:445]: What happens if boclist is the last element in the parent?
would https://www.w3schools.com/jsref/met_node_insertadjacentelement.asp be a better option? (then also for line 429)

@a-ctor

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 25]
```diff
@@ -450,18 +460,18 @@ class BocList
     return false;
   }
 
-  private static FixUpScrolling(bocList: JQuery): void
+  private static FixUpScrolling(bocList: HTMLElement): void
   {
-    var tableBlock = bocList.children('div.bocListTableBlock').first();
+    var tableBlock = bocList.querySelector(':scope>div.bocListTableBlock')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:465]:
```suggestion
    var tableBlock = bocList.querySelector(':scope > div.bocListTableBlock')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 26]
```diff
@@ -450,18 +460,18 @@ class BocList
...
 
     var scrollTimer: Nullable<number> = null;
-    var tableContainer = tableBlock.children('div.bocListTableContainer').first();
-    var scrollableContainer = tableContainer.children('div.bocListTableScrollContainer').first();
+    var tableContainer = tableBlock.querySelector<HTMLElement>(':scope>div.bocListTableContainer')!;
+    var scrollableContainer = tableContainer.querySelector<HTMLElement>(':scope>div.bocListTableScrollContainer')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:469]:
```suggestion
    var tableContainer = tableBlock.querySelector<HTMLElement>(':scope > div.bocListTableContainer')!;
    var scrollableContainer = tableContainer.querySelector<HTMLElement>(':scope > div.bocListTableScrollContainer')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 27]
```diff
@@ -492,116 +502,123 @@ class BocList
...
-    fakeTable.css({ width: '100%' });
-  
-    var realTableHead = table.children('thead').first();
-    var fakeTableHead = realTableHead.clone(true, true);
-    realTableHead.attr({
-      'aria-hidden': 'true',
-      'role': 'none'
+    tableContainer.setAttribute('data-boclist-has-fake-table-head', 'true');
+  
+    var table = scrollableContainer.querySelector(':scope>table')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:524]:
```suggestion
    var table = scrollableContainer.querySelector(':scope > table')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 28]
```diff
@@ -492,116 +502,123 @@ class BocList
...
+  
+    var fakeTable = document.createElement('table');
+    fakeTable.setAttribute('role', 'none');
+    table.classList.forEach(className => { fakeTable.classList.add(className) });
+    fakeTable.setAttribute('cellPadding', '0');
+    fakeTable.setAttribute('cellSpacing', '0');
+    fakeTable.style.width = '100%';
+
+    var realTableHead = table.querySelector(':scope>thead')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:533]:
```suggestion
    var realTableHead = table.querySelector(':scope > thead')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 29]
```diff
@@ -492,116 +502,123 @@ class BocList
...
     // BocList_FixHeaderSize() with timeout needs to be called after setup. This is already taken care of at the call-site.
   }
 
-  private static FixHeaderSize(scrollableContainer: JQuery): void
+  private static FixHeaderSize(scrollableContainer: HTMLElement): void
   {
-    var realTable = scrollableContainer.children('table').first();
-    var realTableWidth = realTable[0].offsetWidth;
-    var previousRealTableWidth = realTable.data("bocListPreviousRealTableWidth");
+    var realTable = scrollableContainer.querySelector<HTMLTableElement>(':scope>table')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:580]:
```suggestion
    var realTable = scrollableContainer.querySelector<HTMLTableElement>(':scope > table')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 30]
```diff
@@ -492,116 +502,123 @@ class BocList
...
-    var realTableHeadRowChildren = realTableHeadRow.children();
+    var realTableHead = realTable.querySelector('thead')!;
+    var realTableHeadRow = realTableHead.children[0];
+    var realTableHeadRowChildren = Array.from(realTableHeadRow.children);
   
-    var fakeTableHeadContainer = scrollableContainer.parent().children('div.bocListFakeTableHead').first();
-    var fakeTableHead = fakeTableHeadContainer.find('thead');
-    var fakeTableHeadRow = fakeTableHead.children().eq(0);
-    var fakeTableHeadRowChildren = fakeTableHeadRow.children();
+    var fakeTableHeadContainer = scrollableContainer.parentNode!.querySelector<HTMLTableElement>(':scope>div.bocListFakeTableHead')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:591]:
```suggestion
    var fakeTableHeadContainer = scrollableContainer.parentNode!.querySelector<HTMLTableElement>(':scope > div.bocListFakeTableHead')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 31]
```diff
@@ -492,116 +502,123 @@ class BocList
...
+    scrollableContainer.style.top = fakeTableHeadContainerHeight + 'px';
+    realTable.style.marginTop = (fakeTableHeadContainerHeight * -1) + 'px';
   }
 
-  private static FixHeaderPosition(tableContainer: JQuery, scrollableContainerJQuery: JQuery): void
+  private static FixHeaderPosition(tableContainer: HTMLElement, scrollableContainer: HTMLElement): void
   {
-    var scrollableContainer = scrollableContainerJQuery[0];
-    var fakeTableHeadContainer = tableContainer.children('div.bocListFakeTableHead').first()[0];
+    var fakeTableHeadContainer = tableContainer.querySelector(':scope>div.bocListFakeTableHead')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:619]:
```suggestion
    var fakeTableHeadContainer = tableContainer.querySelector(':scope > div.bocListFakeTableHead')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 32]
```diff
@@ -492,116 +502,123 @@ class BocList
...
-  
-    var fakeTableHeadWidthContainer = fakeTableHeadContainer.children ('div').first();
-    fakeTableHeadWidthContainer.width(realTableWidth);
-    var fakeTableHeadContainerHeight = fakeTableHeadContainer.height();
-    scrollableContainer.css({ top: fakeTableHeadContainerHeight});
-    realTable.css({ 'margin-top': fakeTableHeadContainerHeight * -1 });
-  
-    fakeTableHeadContainer.show();
+
+    var fakeTableHeadWidthContainer = fakeTableHeadContainer.querySelector<HTMLElement> (':scope>div')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:609]:
```suggestion
    var fakeTableHeadWidthContainer = fakeTableHeadContainer.querySelector<HTMLElement> (':scope > div')!;
```

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts [hunk 33]
```diff
@@ -492,116 +502,123 @@ class BocList
...
+    var previousRealTableWidth = parseInt(realTable.dataset.bocListPreviousRealTableWidth!);
     if (previousRealTableWidth == realTableWidth)
       return;
-    realTable.data("bocListPreviousRealTableWidth", realTableWidth);
+    realTable.dataset.bocListPreviousRealTableWidth = realTableWidth.toString();
   
-    var realTableHead = realTable.eq(0).find('thead');
-    var realTableHeadRow = realTableHead.children().eq(0);
-    var realTableHeadRowChildren = realTableHeadRow.children();
+    var realTableHead = realTable.querySelector('thead')!;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:587]: please review all usages of querySelector and querySelectorAll for missing usage of `:scope`

@a-ctor what's the mgmt summary on using / not-using `:scope` ?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 24 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:445]: It inserts at the end if `null` is supplied as reference node. So if no parent is found it behaves like `appendChild`.

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 33 above)
Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:587]: `.children` vs. `.find`
`:scope > ...` vs. `...`

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 24 above)
Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:445]: Ah, the docs about "Node" don't refer to "node.js".

File: Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts (hunk 33 above)
Comment (MichaelKetting) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:587]: Ah, okay, querySelector and querySelectorAll are always rooted. Missed that one :o

Comment (a-ctor) [reply Remotion/ObjectBinding/Web.ClientScript/Scripts/BocList.ts:587]: They are rooted to the element that you apply the `.querySelector` to. So `element.querySelector` will search for the selector only in the descendants of the elements. But for `.children` you want to search for only children of the element you are applying the selector to, which is available in the css selector as `:scope`.

**Review by MichaelKetting — APPROVED** [review]
